### PR TITLE
Refactor proc macros

### PR DESCRIPTION
**Pull request by rory-ocl**
## Description

This is a rather large refactor of the stylus proc macros with the main goal of improving maintainability, testability, and documentation of these macros. The following changes have been made:

- Operate on AST nodes instead of raw tokens. This allows testing of smaller parts of the macros as unit tests. This also allows some composibility within the implementations.
- Use the visitor pattern in several places to avoid confusing nesting of for-loops.
- Separation of export-abi code using "extention" traits so that it is easy to reason about export-abi when needed, and otherwise ignore it.
- Improved error reporting. The main change here is the ability to report multiple errors when compilation fails, allowing the user to see all errors at once. There is also some small changes with regards to consistency between the macros.
- Unit tests for utilities and macro generation at the AST level. This includes a `assert_ast_eq()` function which can be used to compare two ASTs of the same type. If an error occurs, the ASTs are pretty printed and a diff is output so that the issue can be easily addressed.
- Integration tests in the `stylus-proc/tests` directory. These allow compiled macros to be tested as working correctly. Additionally the `stylus-proc/tests/fail` directory will contain failing integration tests. These all fail to compile and include expected stderr outputs.
- Modify code snippets within docstrings so that they are not ignored, and pass as part of the `cargo test` process.
- A `README.md` which will contain information about developing and maintaining these proc macros.
- Support for `solidity.path.separator` to `rust::path::separator` for custom structs within `sol_interface!` definitions.
- Change the folder structure to have all macro implementations in a `macros` module. Easier to find the macro implementation you are looking for.
- Docstrings within the macro implementations for developer reference.
- Utilities for handing inner attributes including those which contain tokens which need to be parsed.

A little more work is going to be done before this is ready for merge, but it will be good to get this out for review since it is such a large set of changes.

## Checklist

- [ ] I have documented these changes where necessary.
- [ ] I have read the [DCO][DCO] and ensured that these changes comply.
- [ ] I assign this work under its [open source licensing][terms].

[DCO]: https://github.com/OffchainLabs/stylus-sdk-rs/blob/stylus/licenses/DCO.txt
[terms]: https://github.com/OffchainLabs/stylus-sdk-rs/blob/stylus/licenses/COPYRIGHT.md

### CODE REVIEW

File: stylus-proc/src/macros/derive/solidity_error/mod.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            impl From<#ty> for #self_name {
+                fn from(value: #ty) -> Self {
+                    #self_name::#name(value)
+                }
+            }
+        });
+        self.match_arms.push(parse_quote! {
+            #self_name::#name(e) => stylus_sdk::call::MethodError::encode(e),
+        });
+        self._ext.add_variant(field);
```
[the remark this answers is not in the document]
Comment (chrisco512) [reply stylus-proc/src/macros/derive/solidity_error/mod.rs:56]: the warning I get from clippy:

> passing a unit value to a function
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unit_arg
`#[warn(clippy::unit_arg)]` on by defaultclippy[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2FUsers%2Frai%2Fprojects%2Fstylus-sdk-rs%2Fstylus-proc%2Fsrc%2Fmacros%2Fderive%2Fsolidity_error%2Fmod.rs)
mod.rs(56, 9): move the expression in front of the call and replace it with the unit literal `()`: `self._ext;
        ().add_variant(field)`

File: stylus-proc/src/imports.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        pub const SolTypeValue: ConstPath =
+            ConstPath("stylus_sdk::alloy_sol_types::private::SolTypeValue");
+    }
+
+    pub mod sol_data {
+        use syn::parse::Parse;
+
+        use crate::imports::ConstPath;
+
+        pub const Address: ConstPath = ConstPath("stylus_sdk::alloy_sol_types::sol_data::Address");
```
[the remark this answers is not in the document]
Comment (rory-ocl) [reply stylus-proc/src/imports.rs:47]: Right now I just have the imports that were declared as constants before for easy interpolation. Could potentially add more in the future as needed.

**Review by chrisco512 — DISMISSED** [review]

File: stylus-proc/src/macros/derive/solidity_error/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rory-ocl) [reply stylus-proc/src/macros/derive/solidity_error/mod.rs:56]: So the extension here (and in public macro), is designed to do nothing by default. That's why the default impl uses a `()` and returns `Nothing`. When the `export-abi` feature is enabled the `Extension` type alias points to a type that actually provides the logic. This just avoids having `#[cfg(feature = "export-abi")]` in a bunch of different places

File: stylus-sdk/src/abi/mod.rs [hunk 3]
```diff
@@ -98,12 +102,68 @@ macro_rules! function_selector {
...
+}
+
+/// ABI encode a value
+pub fn encode<T>(value: &T) -> Vec<u8>
+where
+    T: AbiType + SolTypeValue<<T as AbiType>::SolType>,
+{
+    T::SolType::abi_encode(value)
+}
+
```
[the remark this answers is not in the document]
Comment (rory-ocl) [reply stylus-sdk/src/abi/mod.rs:121]: I'm not quite sure yet, it wasn't needed for these changes but I do want to take a broader look at how we handle abi encoding/decoding and revisit then.

File: stylus-sdk/src/abi/mod.rs [hunk 4]
```diff
@@ -98,12 +102,68 @@ macro_rules! function_selector {
...
+fn test_encode_decode_params<T, B>(value: T, buffer: B)
+where
+    T: core::fmt::Debug + PartialEq + AbiType + SolTypeValue<<T as AbiType>::SolType>,
+    for<'a> <<T as AbiType>::SolType as SolType>::Token<'a>: TokenSeq<'a>,
+    B: core::fmt::Debug + AsRef<[u8]>,
+{
+    let encoded = encode_params(&value);
+    assert_eq!(encoded, buffer.as_ref());
+    // TODO: debug this assertion for test encode_decode_bytes_tuple()
+    // assert_eq!(abi::encoded_size(&value), encoded.len());
```
[the remark this answers is not in the document]
Comment (rory-ocl) [reply stylus-sdk/src/abi/mod.rs:152]: Not that I am aware of. I haven't had a chance to dig into this but bytes have been working correctly. I'll revisit this with some other abi encoding/decoding changes mentioned above.

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

File: stylus-proc/src/macros/derive/solidity_error/export_abi.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                fn inner_types() -> alloc::vec::Vec<stylus_sdk::abi::export::internal::InnerType> {
+                    use alloc::{format, vec};
+                    use core::any::TypeId;
+                    use stylus_sdk::abi::export::internal::InnerType;
+                    use stylus_sdk::alloy_sol_types::SolError;
+
+                    vec![
+                        #(
+                            InnerType {
+                                name: format!("error {};", <#errors as SolError>::SIGNATURE.replace(',', ", ")),
```
Comment (rauljordan) [inline stylus-proc/src/macros/derive/solidity_error/export_abi.rs:34]: This PR is awesome so far. The only thing I would add is more code comments, such as why this replace is necessary

File: stylus-proc/src/macros/entrypoint.rs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+fn struct_entrypoint_fn(name: &Ident) -> syn::ItemFn {
+    parse_quote! {
+        fn #STRUCT_ENTRYPOINT_FN(input: alloc::vec::Vec<u8>) -> stylus_sdk::ArbResult {
+            use stylus_sdk::{abi::Router, alloy_primitives::U256, console, hex, storage::StorageType};
+            use core::convert::TryInto;
+            use alloc::vec;
+
+            if input.len() < 4 {
+                console!("calldata too short: {}", hex::encode(input));
+                return Err(vec![]);
```
Comment (rauljordan) [inline stylus-proc/src/macros/entrypoint.rs:134]: wouldnt it be better to return a more expressive error here? Maybe for a future pr

File: stylus-proc/src/macros/public/export_abi.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            self_ty,
+            where_clause,
+            inheritance,
+            funcs,
+            ..
+        } = iface;
+
+        let name = match self_ty {
+            syn::Type::Path(path) => path.path.segments.last().unwrap().ident.clone().to_string(),
+            _ => todo!(),
```
Comment (rauljordan) [inline stylus-proc/src/macros/public/export_abi.rs:35]: Why todo? Will this change later on? If so, maybe track with with a code comment or linear task, such as TODO(LINEAR_ITEM_NUMBER)

**Review by rauljordan — COMMENTED** [review]
Fantastic refactor. It will be a lot easier to get into the codebase and extend the sdk than before. Nothing major I could find, although to really go through it with a fine toothed-comb, will need to expand a lot of these macros once more with some sample programs. Awesome job!

**Pull request merged**